### PR TITLE
initial ESP32 support PR

### PR DESCRIPTION
**Pull request by dhalbert**
Initial support for plain ESP32, using Adafruit Feather ESP32 V2 as a starting point. Significant known problems.  @tannewt suggested we go ahead and merge this, since there are many changes we'd like to incorporate even before ESP32 is debugged.

- `*DEBUG_UART*` is now `*CONSOLE_UART*`, many places.
- `CIRCUITPY_CONSOLE_UART` added to specify when the REPL console is done over UART.
- Available support (commented-out by default) in `boards/adafruit_feather_esp32_v2/sdkconfig` for separate ESP logging UART channel on TX and RX pins.
- `espressif/modules` removed, and `mpconfigport.mk` files updated to remove `CIRCUITPY_MODULE` settings. Module variations are now based automatically on `CONFIG_` values from `sdkconfig` setups, including pin never-resets.
- Spaces added around `=` in `CIRCUITPY_ESP_FLASH_* settings in `mpconfigboard.mk`, for uniformity.
- `GPIO<nn>_EXISTS` values defined in `espressif/peripherals/*/pins.c` for use in `common-hicrocontroller/__init__.c` to define only pins that exist (and maybe elsewhere).
- Many additions of `CONFIG_IDF_TARGET_ESP32` in various places, to handle ESP32-only variations.
- `CIRCUITPY_STATUS_BAR` added, to control whether status bar is used. This will probably change later. For right now, it's just to turn off the status bar on the ESP32 V2 so it won't interfere with Thonny, etc.

I did smoke tests on a couple of ESP32S2 and S3 boards (not C3) to make sure their builds aren't completely broken. Only simple manual REPL tests done.

#### Issues

~~Thonny does not work with ESP32. It causes a crash. It may be a gc or other memory problem. So it's hard to get files onto the ESP32. @jepler will be taking a look for now. Some logging help at the crash site:
[emitglue.patch.zip](https://github.com/adafruit/circuitpython/files/9074770/emitglue.patch.zip)~~

Garbage characters are often sent on the REPL port when it first starts up, especially (or only?) after a power-up.

# Testing & Status
Below tests on Adafruit Feather ESP32 V2 in repl in tio, unless otherwise noted
 - [x] thonny: can upload code.py, can interact with repl
 - [x] Neopixel: status shows after program completes
 - [X] AnalogIn/AnalogOut: at repl, jumper betweenn A0/A1. output a ramp on A0, read DAC on A1 
 - [X] DigitalO: At repl, jumper between A0/A1. Switch A0 to output. Set value high & low & read A1.
 - [X] i2c scan: AHT20 stemma qt sensor on stemma qt connector found by busio.I2C.scan
 - [X] i2c device: Uploaded adafruit_ahtx0.py with thonny. Imported at repl and retrieved `aht.temperature` property with plausible value
 - [X] spi loopback: 
   ```
    >>> s = busio.SPI(MOSI=board.MOSI, MISO=board.MISO, clock=board.SCK)
    >>> b = b"x" * 8
    >>> r = bytearray(len(b))
    >>> s.try_lock()
    True
    >>> s.write_readinto(b, r)
    >>> r
    bytearray(b'xxxxxxxx')
    ```
 - [x] Wifi: import, scan, connect. didn't test with adafruit_requests yet. didn't test with wifi workflow yet.
 - [ ] Wifi: can only successfully scan or associate once between hard resets

### CODE REVIEW

**Comment by dhalbert** [comment]
Somehow the frozen libraries and samd-peripherals were way out of date, not sure how. Updated these.

**Comment by dhalbert** [comment]
@jepler you can undraft and merge this whenever you feel it's good enough. As @tannewt suggested merging even with the bugs will help move other things forward.

**Comment by jepler** [comment]
Another scenario that causes misbehavior:
```
>>> s = "x" * 999
>>> import gc
>>> gc.collect()
>>> s
'\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx'
```

this slightly different sequence more consistently causes a guru meditation instead of just printing wrong data:
```
Adafruit CircuitPython 8.0.0-alpha.1-38-g9d77a36fb-dirty on 2022-07-12; Adafruit Feather ESP32 V2 with ESP32
>>> 
>>> dir()
['__name__']
>>> s = "x" * 999
>>> import gc
>>> gc.collect()
>>> s
```

which decodes here as
```
0x400dec57: mp_obj_get_type at /home/jepler/src/circuitpython/ports/espressif/../../py/obj.c:68
 (inlined by) mp_obj_print_helper at /home/jepler/src/circuitpython/ports/espressif/../../py/obj.c:133
 (inlined by) mp_obj_print_helper at /home/jepler/src/circuitpython/ports/espressif/../../py/obj.c:114
0x400e1a25: fun_builtin_1_call at /home/jepler/src/circuitpython/ports/espressif/../../py/objfun.c:75
0x400dd016: mp_call_function_n_kw at /home/jepler/src/circuitpython/ports/espressif/../../py/runtime.c:665
0x400eac99: mp_execute_bytecode at /home/jepler/src/circuitpython/ports/espressif/../../py/vm.c:936
0x400e1ae9: fun_bc_call at /home/jepler/src/circuitpython/ports/espressif/../../py/objfun.c:297 (discriminator 4)
0x400dd016: mp_call_function_n_kw at /home/jepler/src/circuitpython/ports/espressif/../../py/runtime.c:665
0x400dd03a: mp_call_function_0 at /home/jepler/src/circuitpython/ports/espressif/../../py/runtime.c:638
0x40117c03: parse_compile_execute at /home/jepler/src/circuitpython/ports/espressif/../../shared/runtime/pyexec.c:146
0x4011800d: pyexec_friendly_repl at /home/jepler/src/circuitpython/ports/espressif/../../shared/runtime/pyexec.c:734
0x400eeded: run_repl at /home/jepler/src/circuitpython/ports/espressif/../../main.c:823
 (inlined by) main at /home/jepler/src/circuitpython/ports/espressif/../../main.c:922
0x400ef5e3: app_main at /home/jepler/src/circuitpython/ports/espressif/supervisor/port.c:410
0x401bb461: main_task at /home/jepler/src/circuitpython/ports/espressif/build-adafruit_feather_esp32_v2/esp-idf/../../esp-idf/components/freertos/port/port_common.c:141
```
.. an invalid pointer within `mp_obj_get_type`

**Comment by jepler** [comment]
And it doesn't affect just the repl, I a similar snippet in code.py via the repl and it misbehaves in yet a slightly different way:
```
>>> with open("/code.py", "wt") as c: c.write("print(dir())\ns='x'*999\nimport gc\ngc.collect();print(s)\n")
... 
55
>>> 
soft reboot

Auto-reload is on. Simply save files over USB to run them or enter REPL to disable.
code.py output:
['__name__', '__file__']
Traceback (most recent call last):
  File "code.py", line 4, in <module>
NameError: name 's' is not defined
```

**Comment by dhalbert** [comment]
I tried disabling PSRAM completely to see if it was being misused, but that caused a boot loop. I did not have time to figure out what was wrong.

Note that there is a workaround flag for an issue with PSRAM with earlier chip versions. Also I did not check whether there are cache settings that might be wrong.

**Comment by jepler** [comment]
This sequence doesn't involve gc.collect (nor is garbage implicitly collected) and still gives a nonsense result.
```
Adafruit CircuitPython 8.0.0-alpha.1-39-gafbaa2eb6-dirty on 2022-07-13; Adafruit Feather ESP32 V2 with ESP32
>>> import ulab
>>> u = ulab.numpy.ones((6400,))
>>> sum(u)
6.326e+22
```

It crashes at this point instead of returning to the repl:
```
? Backtrace:0x400d4198:0x3ffbf0400x400de76e:0x3ffbf060 0x400e0eed:0x3ffbf090 0x400e3495:0x3ffbf0b0 0x400e9d8a:0x3ffbf0d0 0x400e19dd:0x3ffbf100 0x400dcfce:0x3ffbf120 0x400eac51:0x3ffbf140 0x400e1aa1:0x3ffbf1e0 0x400dcfce:0x3ffbf210 0x400dcff2:0x3ffbf230 0x40117bc7:0x3ffbf250 0x40117fd1:0x3ffbf2f0 0x400eeda5:0x3ffbf390 0x400ef573:0x3ffbf400 0x401bb431:0x3ffbf420 
0x400d4198: find_qstr at /home/jepler/src/circuitpython/ports/espressif/../../py/qstr.c:144
 (inlined by) qstr_hash at /home/jepler/src/circuitpython/ports/espressif/../../py/qstr.c:288
0x400e0eed: mp_obj_dict_store at /home/jepler/src/circuitpython/ports/espressif/../../py/objdict.c:650
0x400e3495: module_attr at /home/jepler/src/circuitpython/ports/espressif/../../py/objmodule.c:111
0x400e9d8a: mp_builtin___repl_print__ at /home/jepler/src/circuitpython/ports/espressif/../../py/modbuiltins.c:464
 (inlined by) mp_builtin___repl_print__ at /home/jepler/src/circuitpython/ports/espressif/../../py/modbuiltins.c:457
0x400e19dd: fun_builtin_1_call at /home/jepler/src/circuitpython/ports/espressif/../../py/objfun.c:75
0x400dcfce: mp_call_function_n_kw at /home/jepler/src/circuitpython/ports/espressif/../../py/runtime.c:665
0x400eac51: mp_execute_bytecode at /home/jepler/src/circuitpython/ports/espressif/../../py/vm.c:936
0x400e1aa1: fun_bc_call at /home/jepler/src/circuitpython/ports/espressif/../../py/objfun.c:297 (discriminator 4)
0x400dcfce: mp_call_function_n_kw at /home/jepler/src/circuitpython/ports/espressif/../../py/runtime.c:665
0x400dcff2: mp_call_function_0 at /home/jepler/src/circuitpython/ports/espressif/../../py/runtime.c:638
0x40117bc7: parse_compile_execute at /home/jepler/src/circuitpython/ports/espressif/../../shared/runtime/pyexec.c:146
0x40117fd1: pyexec_friendly_repl at /home/jepler/src/circuitpython/ports/espressif/../../shared/runtime/pyexec.c:734
0x400eeda5: run_repl at /home/jepler/src/circuitpython/ports/espressif/../../main.c:823
 (inlined by) main at /home/jepler/src/circuitpython/ports/espressif/../../main.c:925
0x400ef573: app_main at /home/jepler/src/circuitpython/ports/espressif/supervisor/port.c:410
0x401bb431: main_task at /home/jepler/src/circuitpython/ports/espressif/build-adafruit_feather_esp32_v2/esp-idf/../../esp-idf/components/freertos/port/port_common.c:141
```

**Comment by jepler** [comment]
I think I found the problem with psram. Now, thonny works with the adafruit feather esp32 v2!

image

File: ports/espressif/supervisor/port.c
Comment (tannewt) [inline ports/espressif/supervisor/port.c:109]: Want to use 176 like the S3? Not 48 like the S2.

File: ports/espressif/supervisor/port.c
Comment (tannewt) [inline ports/espressif/supervisor/port.c:223]: Does spiram_size_usable_for_malloc require you to let the IDF manage the spiram? Generally we have it ignore it.

**Review by tannewt — COMMENTED** [review]
Two minor things we may want to change before merge. Nice work @dhalbert and @jepler!

**Comment by jepler** [comment]
Wifi scanning/connecting only works once, then you have to use the reset button.

File: ports/espressif/supervisor/port.c
Comment (jepler) [reply ports/espressif/supervisor/port.c:109]: allocating 176kB fails, which is why I changed it.

File: ports/espressif/supervisor/port.c
Comment (jepler) [reply ports/espressif/supervisor/port.c]: I renamed the function to avoid confusion. It has nothing to do with malloc, it's just the name of the function I copied in from esp-idf, where it was defined statically.

**Comment by jepler** [comment]
The main problem i'm aware of at this point is that wifi can only scan or associate once before a hard-reset.

**Review by tannewt — DISMISSED** [review]
Ok, sounds good. We should be aware that the heap on non-PSRAM boards may actually shrink if we were previously giving them more than half of internal ram.

**Comment by jepler** [comment]
I'll merge this with green; if I have additional fixes (I'm chasing the wifi 'only once' problem next) I'll start fresh PRs. Re-starting the full CI on this PR is getting a little old.

**Review by tannewt — APPROVED** [review]
Thanks again!

**Comment by tannewt** [comment]
One remaining check is hung uploading an artifact. Will merge anyway.

**Pull request merged**
